Add style guide (#3300)
fishtown-analytics/docs.getdbt.com: website/docs/guides/best-practices/how-we-style/1-how-we-style-our-dbt-models.md

---
title: How we style our dbt models
id: 1-how-we-style-our-dbt-models
---

> ## Fields and model names

Reviewer: For model naming, I did like the old "All objects should be plural" from our previous style guide. I've had customers run into snags a couple of times in RO's because they weren't sure if the model name was, e.g. `stg_payment` or `stg_payments`, so I think it's a good thing to standardize.

Reviewer: Another good one from the old guide is limiting abbreviations in model and column names for readability.

- Each model should have a primary key.

> - The primary key of a model should be named `<object>_id`, e.g. `account_id` – this makes it easier to know what `id` is being referenced in downstream joined models.

Reviewer: Following this line of thought, I feel like it would be helpful to add another line about the data type for primary keys so developers don't end up trying to join columns with inconsistent data types. Something like "The primary key of a model should be a string"

- Booleans should be prefixed with `is_` or `has_`.

> - Timestamp columns should be named `<event>_at`, e.g. `created_at`, and should be in UTC. If a different timezone is being used, this should be indicated with a suffix, e.g `created_at_pt`.

Reviewer: Should dates use the same naming convention?

- Price/revenue fields should be in decimal currency (e.g. `19.99` for $19.99; many app databases store prices as integers in cents). If non-decimal currency is used, indicate this with suffix, e.g. `price_in_cents`.

> - Avoid reserved words as column names

Reviewer: ```suggestion
- Avoid reserved words as column names.
```

---
title: How we style our dbt models
id: 1-how-we-style-our-dbt-models
---

## Fields and model names

> - 👥 Models should be pluralized, e.g. `customers`, `orders`, `products`.

Reviewer 2: ```suggestion
- 👥 Models should be pluralized, for example, `customers`, `orders`, `products`.
```

> - 🔑 Each model should have a primary key.

Reviewer 3: Do you have opinions about whether this is a single column (implying `dbt_utils.generate_surrogate_key(['region_id', 'user_id'])`) or are you OK with the two columns `region_id` + `user_id` fulfilling this duty between them?

> - 🔑 The primary key of a model should be named `<object>_id`, e.g. `account_id` – this makes it easier to know what `id` is being referenced in downstream joined models.

Reviewer 2: ```suggestion
- 🔑 The primary key of a model should be named `<object>_id`, for example, `account_id`. This makes it easier to know what `id` is being referenced in downstream joined models.
```

- 🔑 Keys should be string data types.

> - 🔑 Consistency is key! Use the same field names across models where possible, e.g. a key to the `customers` table should be named `customer_id` rather than `user_id` or 'id'.

Reviewer 2: ```suggestion
- 🔑 Consistency is key! Use the same field names across models where possible. For example, a key to the `customers` table should be named `customer_id` rather than `user_id` or 'id'.
```

> - ❌ Do not use abbreviations or aliases, emphasize readability over brevity, e.g. `cust` for `customer` or `o` for `orders`.

Reviewer 2: ```suggestion
- ❌ Do not use abbreviations or aliases. Emphasize readability over brevity. For example, do not use `cust` for `customer` or `o` for `orders`.
```

- ❌ Avoid reserved words as column names.
- ➕ Booleans should be prefixed with `is_` or `has_`.

> - 🕰️ Timestamp columns should be named `<event>_at`, e.g. `created_at`, and should be in UTC. If a different timezone is being used, this should be indicated with a suffix, e.g `created_at_pt`.

Reviewer 2: ```suggestion
- 🕰️ Timestamp columns should be named `<event>_at`(for example, `created_at`) and should be in UTC. If a different timezone is used, this should be indicated with a suffix (`created_at_pt`).
```

> - 📆 Dates should be named `<event>_date`, e.g. `created_date`.

Reviewer 2: ```suggestion
- 📆 Dates should be named `<event>_date`. For example, `created_date.`
```

> - 🔙 Events dates and times should be past tense, e.g. `created`, `updated`, `deleted`.

Reviewer 2: ```suggestion
- 🔙 Events dates and times should be past tense &mdash; `created`, `updated`, or `deleted`.
```

> - 💱 Price/revenue fields should be in decimal currency (e.g. `19.99` for $19.99; many app databases store prices as integers in cents). If non-decimal currency is used, indicate this with suffix, e.g. `price_in_cents`.

Reviewer 2: ```suggestion
- 💱 Price/revenue fields should be in decimal currency (`19.99` for $19.99; many app databases store prices as integers in cents). If a non-decimal currency is used, indicate this with a suffix (`price_in_cents`).
```

- 🐍 Schema, table and column names should be in `snake_case`.

> - 🏦 Use names based on the _business_ terminology, rather than the source terminology. For example, if the source database uses `user_id` but the business calls them `customer_id`, use `customer_id` in the model.

Reviewer 3: All the way through, or only as the models get closer to being marts? I saw a cool slide from a meetup which I don't know is for public consumption but I will DM you

Author: i let this marinate overnight, at first i was like 'let's add some nuance to this and soften it' but honestly, the more i thought about it the more i was like 'you're going to get a *lot* more clarity out of using unified terminology and it feels like a core function of AE to define and implement that' - i can't think of a compelling case for not immediately shifting to unified business terms at the staging layer except for exceptions where you have multiple source systems that need to be unified and have an underlying unification layer.

> - 🔢 Versions of models should use the suffix `_v1`, `_v2`, etc for consistency, e.g. `customers_v1`, `customers_v2`.

Reviewer 2: ```suggestion
- 🔢 Versions of models should use the suffix `_v1`, `_v2`, etc for consistency (`customers_v1` and `customers_v2`).
```

> - 🗄️ Use a consistent ordering of data types and consider grouping and labeling columns by type as in the exampe below. This will minimize join errors and make it easier to read the model, as well as help downstream consumers of the data understand the data types and scan models for the columns they need. We prefer to use the following order: ids, strings, numerics, booleans, dates, timestamps.

Reviewer 2: ```suggestion
- 🗄️ Use a consistent ordering of data types and consider grouping and labeling columns by type, as in the example below. This will minimize join errors and make it easier to read the model, as well as help downstream consumers of the data understand the data types and scan models for the columns they need. We prefer to use the following order: ids, strings, numerics, booleans, dates, and timestamps.
```
